### PR TITLE
Fix/outgoing deletes

### LINKED ISSUE #476: Announce Delete
If a magazine receives a delete activity from a remote instance it has to announce it

### PR DESCRIPTION
**Pull request by BentiGorlich**
### Deleting a post will now always lead to an outgoing delete message if:

- the entry is local **or**
- the magazine is local (announce) **or**
- the actor is a moderator in the magazine (announce)
- **AND**
- the author of the post and the actor of the delete are from the same server **or**
- the actor is a moderator in the magazine

### Remove the admin as the "owner" of all remote magazines

- the magazine will not be created with the admin as the owner anymore
- the owner of the magazine is not excluded from being removed when updating a remote magazine anymore

### Minor fixes

- Moderators Collections -> Context is now an array like it should be

Closes #476, #529

### CODE REVIEW

**Comment by melroy89** [comment]
> * the magazine will not be created with the admin as the owner anymore

So I suspect the creator will become the owner of the new magazine, correct?

**Comment by nobodyatroot** [comment]
Is there something that needs to change for the mod panel in the sidebar? I see I show up now in mbinreleases but not in the side panel widget:

https://kbin.run/m/mbinReleases@gehirneimer.de

image

https://kbin.run/m/mbinReleases@gehirneimer.de/moderators

image

File: src/Service/ActivityPub/Wrapper/DeleteWrapper.php [hunk 1]
```diff
@@ -28,15 +33,25 @@ public function __construct(
...
     ])]
-    public function build(ActivityPubActivityInterface $item, string $id): array
+    public function build(ActivityPubActivityInterface $item, string $id, User $deletingUser = null): array
     {
         $item = $this->factory->create($item);
 
+        $userUrl = $item['attributedTo'];
+
+        if (null !== $deletingUser) {
+            if (null !== $deletingUser->apId) {
```
Comment (melroy89) [inline src/Service/ActivityPub/Wrapper/DeleteWrapper.php:43]: Maybe add some comments for future reference, including for yourself ;)

**Comment by BentiGorlich** [comment]
> > * the magazine will not be created with the admin as the owner anymore
> 
> So I suspect the creator will become the owner of the new magazine, correct?

Its been this way before as well. What changed is that remote mags are not created with the admin as owner. From now on remote mags just don't have an owner anymore, because they do not need one

File: src/Service/ActivityPub/Wrapper/DeleteWrapper.php (hunk 1 above)
Comment (BentiGorlich) [reply src/Service/ActivityPub/Wrapper/DeleteWrapper.php:43]: true

**Comment by BentiGorlich** [comment]
> Is there something that needs to change for the mod panel in the sidebar? I see I show up now in mbinreleases moderator page but not in the side panel widget:
> 
> https://kbin.run/m/mbinReleases@gehirneimer.de
> 
> image
> 
> https://kbin.run/m/mbinReleases@gehirneimer.de/moderators
> 
> image

I suspect there is code that is not needed anymore, but my guess is that your instancedid not update the magazine yet and therefore the admin is still the "owner" of the magazine. I wondered if I should write a migration that deletes all moderator entries where the apId of the mag is not null and the moderator entry says "isOwner"

**Review by nobodyatroot — APPROVED** [review]
LGTM, I assume this fixes @TheVillageGuy's delete problem? Wasn't sure since I don't see an issue for it. Migration worked as expected and after interacting with mbinReleases@gehirneimer.de I show up in the mod panel now.

**Comment by TheVillageGuy** [comment]
I'll check it out soon

**Comment by melroy89** [comment]
> > So I suspect the creator will become the owner of the new magazine, correct?
> 
> Its been this way before as well. What changed is that remote mags are not created with the admin as owner. From now on remote mags just don't have an owner anymore, because they do not need one

ah of course. Great!

File: src/Service/ActivityPub/Wrapper/DeleteWrapper.php (hunk 1 above)
Comment (melroy89) [reply src/Service/ActivityPub/Wrapper/DeleteWrapper.php:44]: But you didn't add comments

**Review by melroy89 — APPROVED** [review]

**Comment by TheVillageGuy** [comment]
I have tried deleting https://gehirneimer.de/m/RimWorldPorn@rimworld.gallery/t/227289/Anomaly-Spoilers-Post-ending-screenshot by moderation again and as you can tell it's still there, so my problem isn't fixed

**Comment by BentiGorlich** [comment]
Do you have this PR running on your instance? otherwise it will not work

**Comment by TheVillageGuy** [comment]
Oh, no I don't I thought it had already been merged.

**Comment by TheVillageGuy** [comment]
It works for me

**Pull request merged**